### PR TITLE
Upstream merge 2024 11 18

### PR DESCRIPTION
**Pull request by skmcgrail**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: ssl/handshake_client.cc [hunk 1]
```diff
@@ -1379,7 +1379,33 @@ static enum ssl_hs_wait_t do_send_client_certificate(SSL_HANDSHAKE *hs) {
...
+      if (std::find(hs->certificate_types.begin(), hs->certificate_types.end(),
+                    cert_type) == hs->certificate_types.end()) {
+        OPENSSL_PUT_ERROR(SSL, SSL_R_UNKNOWN_CERTIFICATE_TYPE);
+        return ssl_hs_error;
+      }
+    }
+  } else {
     // Without a client certificate, the handshake buffer may be released.
     hs->transcript.FreeBuffer();
   }
```
Comment (nebeid) [inline ssl/handshake_client.cc:1411]: I understand that the check on l.1413 was already performed on l.1382. I don't see something that changes `hs` in between, right? Do we want to keep just the check `ssl_output_cert_chain`?

File: ssl/internal.h [hunk 2]
```diff
@@ -753,7 +753,7 @@ size_t ssl_cipher_get_record_split_len(const SSL_CIPHER *cipher);
 // AES in hardware is available.
```
Comment (nebeid) [inline ssl/internal.h:753]: |group_id| is to be removed from the documentation above as well.

**Review by justsmth — APPROVED** [review]

**Review by nebeid — DISMISSED** [review]

File: ssl/handshake_client.cc [hunk 3]
```diff
@@ -1379,17 +1379,41 @@ static enum ssl_hs_wait_t do_send_client_certificate(SSL_HANDSHAKE *hs) {
     }
   }
 
-  if (!ssl_has_certificate(hs)) {
+  if (!ssl_on_certificate_selected(hs)) {
+      return ssl_hs_error;
```
Comment (nebeid) [inline ssl/handshake_client.cc:1383]: indent

File: ssl/internal.h [hunk 4]
```diff
@@ -747,13 +747,13 @@ bool ssl_cipher_requires_server_key_exchange(const SSL_CIPHER *cipher);
 size_t ssl_cipher_get_record_split_len(const SSL_CIPHER *cipher);
 
 // ssl_choose_tls13_cipher returns an |SSL_CIPHER| corresponding with the best
-// available from |client_cipher_suites| compatible with |version|, |group_id| and
+// available from |client_cipher_suites| compatible with |version|, and
```
Comment (nebeid) [inline ssl/internal.h:750]: comma not needed

**Review by nebeid — COMMENTED** [review]
I added some nit comments.

File: include/openssl/bytestring.h [hunk 5]
```diff
@@ -58,17 +58,20 @@ struct cbs_st {
 
 // CBS_init sets |cbs| to point to |data|. It does not take ownership of
 // |data|.
-OPENSSL_EXPORT void CBS_init(CBS *cbs, const uint8_t *data, size_t len);
+OPENSSL_INLINE void CBS_init(CBS *cbs, const uint8_t *data, size_t len) {
+  cbs->data = data;
+  cbs->len = len;
+}
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply include/openssl/bytestring.h:64]: Going to nuke the commit

**Review by nebeid — APPROVED** [review]

**Review by justsmth — APPROVED** [review]

**Pull request merged**